ztp/source-crs-builder/MachineConfigChronyDynamicMaster/ptp-sync-check

```diff
@@ -0,0 +1,157 @@
+  if ! var_defined $LOG_DIR; then
+    return 1
+  fi
+  LAST_LOG_STR=$(ls -ltr $LOG_DIR/ |tail -1)
+  if ! var_defined $LAST_LOG_STR; then
+    return 1
+  fi
+  LAST_LOG_ARRAY=($LAST_LOG_STR)
+  LOG_NAME=${LAST_LOG_ARRAY[-1]}
+  if ! var_defined $LOG_NAME; then
+    return 1
+  fi
+  LAST_STATUS=$(tail -2000 $LOG_DIR/$LOG_NAME |grep "] master offset" |tail -1)
+  return 0
+}
+
+# Get the time offset and associated timestamp from ptp4l log line
+# ARGUMENTS:
+#   none
+# GLOBAL:
+#   LAST_STATUS - A log line containing offset and timestamp
+#     Example: "2021-10-18T16:47:06.022983665+00:00 stdout F ptp4l[8180.150]: [ens4f0] 
+#     master offset 3 s2 freq -16756169 path delay 1481"
+#   SYNC_AGE - Maximum allowed age in seconds of a log line to be considered valid
+#     for sync evaluation
+# OUTPUT:
+#   OFFSET_VALUE - in ns
+# RETURN:
+#   0 if time offset is valid, 1 if not
+get_offset_and_timestamp() {
+  local STRING_ARRAY=($LAST_STATUS)
+  local OFFSET_DATE=${STRING_ARRAY[0]}
+  
+  declare -ig OFFSET_VALUE
+  OFFSET_VALUE=${STRING_ARRAY[7]}
+  declare -ig ABS_OFFSET_VALUE
+  export ABS_OFFSET_VALUE=$(abs $OFFSET_VALUE)
+  log_debug "PTP offset: $OFFSET_VALUE, abs offset $ABS_OFFSET_VALUE,timestamp $OFFSET_DATE"
+
+  # Check how fresh is the offset measurement to protect
+  # from looking at the old log that has not been cleaned from the system
+  local CURRENT_TIME="$(date +%s.%N)" # current date in nanoseconds
+  local OFFSET_TIMESTAMP="$(date +%s.%N -d "$OFFSET_DATE")" # measurement date in ns
+  local TIME_DIFF=$(awk '{print $1-$2}' <<<"${CURRENT_TIME} ${OFFSET_TIMESTAMP}")
+
+  if $(greater $TIME_DIFF $SYNC_AGE); then
+    return 1
+  else
+    return 0
+  fi
+}
+
+# Check if PTP is synced
+# ARGUMENTS:
+#   none
+# GLOBAL:
+#   SYNC_MARKER_PATH - File for starting chronyd conditionally. Used to find the 
+#      last known ptp sync state and apply hysteresis threshold
+#   MAX_OFFSET - Maximum allowed clock offset in nanoseconds
+#   OFFSET_THRESHOLD - hysteresis threshold
+#   ABS_OFFSET_VALUE - PTP clock offset
+# OUTPUT:
+#   none
+# RETURN:
+#   0 if the offset is within the boundaries, 1 if not
+ptp_synced() {
+  if [[ -f "$SYNC_MARKER_PATH" ]]; then
+    log_debug "Last known ptp state - synchronized"
+    SYNC_STATE="Synchronized"
+    ((MAX_OFFSET=MAX_OFFSET+OFFSET_THRESHOLD))
+  else
+    log_debug "Last known ptp state - unsynchronized"
+    SYNC_STATE="Unsynchronized"
+    ((MAX_OFFSET=MAX_OFFSET-OFFSET_THRESHOLD))
+  fi
+  if [ $ABS_OFFSET_VALUE -gt $MAX_OFFSET ]; then
+    return 1
+  fi
+  return 0
+}
+
+
+# If ptp is synced and the offset is valid, create chronyd marker
+# Otherwise delete the marker
+main() {
+  if get_last_ptp_status && get_offset_and_timestamp && ptp_synced; then
+    log_debug "ensure $SYNC_MARKER_PATH exists"
+    touch $SYNC_MARKER_PATH
+  else
+    log_debug "ensure $SYNC_MARKER_PATH does not exist"
+    rm $SYNC_MARKER_PATH &> /dev/nul
+  fi
+}
+
+if [[ "${BASH_SOURCE[0]}" = "${0}" ]]; then
```

Comment:
Modulino++!!

Now... Can I encourage you to add some unit tests for some of the more critical functions?

Reply:
Can I opt for doing it in a separate PR? This one is hanging too long...

Reply:
> Can I opt for doing it in a separate PR? This one is hanging too long...

I'm of the opinion that unit tests should always be in the same commit as the related code, or at least in the same PR, so our CI can prove the new code is okay and does what it intends.

That said, with the changeover to crictl, a lot of the complexity has been removed, and with a lot of our scripts mocking complex cmdline utilities is more trouble that it's worth. I think if you focused your unit testing on get_ptp_offset I'd be happy enough.